src/hotspot/os/bsd/memMapPrinter_macosx.cpp

```diff
@@ -0,0 +1,349 @@
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#if defined(__APPLE__)
+
+#include "precompiled.hpp"
+
+#include "nmt/memMapPrinter.hpp"
+#include "runtime/os.hpp"
+#include "utilities/align.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/powerOfTwo.hpp"
+
+#include <limits.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <libproc.h>
+#include <unistd.h>
+
+#include <mach/vm_inherit.h>
+#include <mach/vm_prot.h>
+#include <mach/mach_vm.h>
+
+/* maximum number of mapping records returned */
+static const int MAX_REGIONS_RETURNED = 1000000;
+
+class MappingInfo {
+public:
+  const char* _address;
+  size_t _size;
+  stringStream _share_buffer;
+  stringStream _type_buffer;
+  stringStream _protect_buffer;
+  stringStream _file_name;
+  const char* _tag_text;
+
+  MappingInfo() {}
+
+  void reset() {
+    _share_buffer.reset();
+    _protect_buffer.reset();
+    _type_buffer.reset();
+    _file_name.reset();
+    _tag_text = "";
+  }
+
+  void process(const proc_regionwithpathinfo& mem_info) {
+    reset();
+
+    const proc_regioninfo& rinfo = mem_info.prp_prinfo;
+
+    _address = (const char*) rinfo.pri_address;
+    _size = rinfo.pri_size;
+
+    if (mem_info.prp_vip.vip_path[0] != '\0') {
+      _file_name.print_raw(mem_info.prp_vip.vip_path);
+    }
+    // proc_regionfilename() seems to give bad results, so we don't use it here.
+
+    char prot[4];
+    char maxprot[4];
+    rwbits(rinfo.pri_protection, prot);
+    rwbits(rinfo.pri_max_protection, maxprot);
+    _protect_buffer.print("%s/%s", prot, maxprot);
+
+    get_share_mode(_share_buffer, rinfo);
+    _tag_text = tagToStr(rinfo.pri_user_tag);
+  }
+
+  static void get_share_mode(outputStream& out, const proc_regioninfo& rinfo) {
+    static const char* share_strings[] = {
+      "cow", "pvt", "---", "shr", "tsh", "p/a", "s/a", "lpg"
+    };
+    const bool valid_share_mode = rinfo.pri_share_mode >= SM_COW && rinfo.pri_share_mode <= SM_LARGE_PAGE;
+    if (valid_share_mode) {
+      int share_mode = rinfo.pri_share_mode;
+      out.print_raw(share_strings[share_mode - 1]);
```

Comment:
this makes me nervous. Please do a bounds check beforehand. Assert is fine. Who knows if Apple changes the numeric definitions for SM_xxx

Reply:
Done.